### PR TITLE
Added cleanup task to start of every execute.

### PR DESCRIPTION
**Pull request by Jeffery-Wasty**
Under certain conditions, prepared statement cache can end up being filled quicker than expected, while being cleaned less often than expected. This adds additional cache clean triggers making the cleaning much more frequent.

### CODE REVIEW

**Comment by ecki** [comment]
Do you really want to force=true the cleanup, doesn’t that add roundtrips?

**Comment by Jeffery-Wasty** [comment]
For sure this can be improved. I have only opened this PR so the user can see our proposed changes and try testing them out, and see if that resolves the issue. For an actual fix, we will look into a cleaner and more performance-friendly solution.

**Comment by ecki** [comment]
> can see our proposed changes and try testing them out, and see if that resolves the issue.

ah yes, I understand, makes sense.

it will check if it fixes the handles accumulated in my reproducer test, but of course not sure if that is the exact problem of OP

**Comment by Jeffery-Wasty** [comment]
No we're not, but from the information we have, this is the best guess.

**Comment by David-Engel** [comment]
For the production fix, I'd set force to false. (Assuming this is the fix.) We don't need to be super aggressive on the cleanup. It's just that if an application is doing millions of executions of a PreparedStatement and some of them occasionally result in an invalidated prepared statement handle, we should clean up periodically to avoid continuously increasing memory usage. Current cleanup only happens on close of the PreparedStatement and connection close.

File: src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/PreparedStatementTest.java [hunk 1]
```diff
@@ -400,9 +398,7 @@ public void testStatementPoolingEviction() throws SQLException {
                 }
 
                 // If we use it, now discard queue should be "full".
-                if (0 == testNo)
-                    assertSame(discardedStatementCount, con.getDiscardedServerPreparedStatementCount());
-                else
+                if (0 != testNo)
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/unit/statement/PreparedStatementTest.java:401]: the test case for testNo==0 was removed we're not testing this anymore?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java [hunk 2]
```diff
@@ -2199,6 +2201,7 @@ public int[] executeBatch() throws SQLServerException, BatchUpdateException, SQL
             loggerExternal.finer(toString() + ACTIVITY_ID + ActivityCorrelator.getCurrent().toString());
         }
         checkClosed();
+        connection.unprepareUnreferencedPreparedStatementHandles(false);
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java:2204]: there are more occurrences of executeStatement in the file should we be doing this for every one of them?

**Review by lilgreenbird — APPROVED** [review]

**Review by barryw-mssql — APPROVED** [review]

**Pull request merged**